syft/messaging/plan.py

```diff
@@ -33,62 +33,134 @@ class func2plan(object):
     This class should be used only as decorator.
     """
 
-    def __init__(self, args_shape=None, verbose=False):
+    def __init__(self, args_shape=None, state=None, verbose=False):
         self.args_shape = args_shape
         self.verbose = verbose
+        if state is not None:
+            self.state = state
+            self.include_state = True
+        else:
+            self.state = {}
+            self.include_state = False
 
     def __call__(self, plan_blueprint):
         plan = Plan(
             owner=sy.local_worker,
             id=sy.ID_PROVIDER.pop(),
             name=plan_blueprint.__name__,
             blueprint=plan_blueprint,
+            state=self.state,
+            include_state=self.include_state,
             verbose=self.verbose,
         )
         if self.args_shape:
             plan._auto_build(args_shape=self.args_shape)
         return plan
 
 
-def method2plan(plan_blueprint):
-    """Converts a method to a plan.
-    Converts a method containing sequential pytorch code into
-    a plan object which can be sent to any arbitrary worker.
-    """
-    plan = Plan(
-        owner=sy.local_worker,
-        id=sy.ID_PROVIDER.pop(),
-        blueprint=plan_blueprint,
-        name=plan_blueprint.__name__,
-        is_method=True,
+def method2plan(*args, **kwargs):
+    raise SyntaxError(
+        "method2plan is not supported anymore. Consider instead subclassing your object from sy.Plan."
     )
 
-    @property
-    def method(self: object) -> Plan:
+
+class State(object):
+    """The State is a Plan attribute and is used to send tensors along functions
+
+    It references Plan tensor or parameters attributes using their name, and make
+    sure they are provided to remote workers who are sent the Plan.
+    """
+
+    def __init__(self, plan):
+        self.keys = []
+        self.plan = plan
+
+    def __repr__(self):
+        return "State: " + ", ".join(self.keys)
+
+    def append(self, key):
```

Comment:
Can you add more details to this docstring? There's a lot going on here, not sure if I follow.